### ISSUE TITLE
Kitura Server on iOS Device

### ISSUE THREAD

**Issue by hactar** [IBM-Swift/Kitura#1229]
## Context and Description
Is https://github.com/IBM-Swift/Kitura-HelloWorld-iOS really the simplest way to run Kitura as a server on an iOS device? Can no cocoapod be provided?

I'm currently looking for options, and while kitura would be my top choice as I love it on linux, it appears to be painful to add when compared with other options such as "swifter" or "embassy" which can be added by adding a single line to my cocoapods file...

**Comment by seabaylea** [comment]
It _should_ be possible to build a CocoaPod using the output on the Builder/Makefile script and either publish that or store it in a separate branch.

**Comment by nodeful** [comment]
I also would like Kitura to be available through CocoaPods as I need to run in as part of my client-side macOS app

**Comment by ianpartridge** [comment]
@ShihabMehboob has been investigating this and will provide some details of our current thoughts soon.

**Comment by ShihabMehboob** [comment]
Hey!

You, along with many users on the public forums had expressed a desire for a standalone [Kitura](https://github.com/IBM-Swift/Kitura) Pod which could be added simply into their existing iOS projects in their Pod files via one simple line (e.g. `pod 'Kitura', '~> 1.0'`), which would substantially simplify the setup process on the user's end, eliminating the need for juggling around with messy imports, and saving them time.

This was investigated by creating podspec files for all Kitura repos and its dependencies, allowing the creation of Pods. The lower level repositories contained within Kitura were the starting point for this, as they contain no dependencies of their own. This included the following:

- Socket ([Repo containing the podspec](https://github.com/IBM-Swift/BlueSocket))
- SSLService ([Repo containing the podspec](https://github.com/IBM-Swift/BlueSSLService))
- LoggerAPI ([PR containing the podspec](https://github.com/IBM-Swift/LoggerAPI/pull/28))
- Kitura-TemplateEngine ([PR containing the podspec](https://github.com/IBM-Swift/Kitura-TemplateEngine/pull/25))
- CCurl ([PR containing the podspec](https://github.com/IBM-Swift/CCurl/pull/17) - this podspec only contains one .h file - does this even need to be a standalone Pod?)

The Pods would be generated by uploading to the Specs repo via `pod trunk push something.podspec`. A typical podspec may look as follows:

> Pod::Spec.new do |s|
>   s.name        = "KituraNet"
>   s.version     = "2.0.4"
>   s.summary     = "The Kitura-net module contains logic for sending and receiving HTTP requests"
>   s.homepage    = "https://github.com/IBM-Swift/Kitura-net"
>   s.license     = { :type => "Apache License, Version 2.0" }
>   s.author     = "IBM"
>   s.module_name  = 'KituraNet'
>   s.requires_arc = true
>   s.osx.deployment_target = "10.11"
>   s.ios.deployment_target = "10.0"
>   s.tvos.deployment_target = "10.0"
>   s.source   = { :git => "https://github.com/IBM-Swift/Kitura-net.git", :tag => s.version }
>   s.source_files = "Sources/KituraNet/*.swift"
>   s.pod_target_xcconfig =  {
>         'SWIFT_VERSION' => '4.0.3',
>   }
> end

As the podspecs rely on including the version number as a tag (e.g. `s.source   = { :git => "https://github.com/IBM-Swift/Kitura-net.git", :tag => s.version }`), the concern of keeping the repos up to date and in sync was raised. If a newer version was to be released of a repo, the podspecs would also have to be updated in accordance to reflect this, which could turn into an unnecessarily laborious task (as there are many new versions often tagged and released of these repos).

This also led into the issue of attempting to create a Pod for a much bigger repo, such as Kitura, which relies upon the other previously mentioned repos, along with Kitura-net too. Two options were introduced at this point:

1. We could create one individual standalone Pod using the raw files from each and every repo within Kitura, manually added into one folder on a pod branch. This would be a messy solution where all the files would have to be gathered from every dependency, and put into one containing folder from which the Pod would be created. The `import` statements above each file would also have to be removed. This is the manner in which KituraKit approached this, as evident [here](https://github.com/IBM-Swift/KituraKit/tree/pod), which required a separate branch to house it, and the source referencing it (e.g. `s.source = { :git => "https://github.com/IBM-Swift/KituraKit", :branch => "pod", :submodules => true }`). You can imagine how large this could grow if the entirety of Kitura were to take on this approach, and how laborious it would be to change every changed file every time something in its dependency tree were to get an update. Another sample of this can be seen [here](https://github.com/ShihabMehboob/KituraContracts/tree/issue.pods) for KituraContracts, which has rearranged the Source folder to include LoggerAPI as the podspec validation complains about missing LoggerAPI if the folder isn't present and rather just attempted to be imported. This has been done as an experiment, and would be far too complex and messy for anything larger such as Kitura.

2. We could carry on the route we did for the previous repos, where a Pod was created for each individual smaller repo and then referenced in the podspecs of the repos that rely on it (e.g. `s.dependency 'BlueSocket', '~> 1.0.2'`). This depends upon all the dependency pods to be tagged correctly, and kept up to date with each version release. This might be fine if something was developed once and never touched again, but not so much when code is ever changing. Whilst this method does not bring in and duplicate every file from its dependencies like the previous option, it relies on adding another Pod as a dependency within a Pod, and each Pod to be up to date and correctly referenced with the version.

Both options had one major issue, the files would have to continually be changed after every release of every dependency that got changed, with either versions correctly tagged and kept in sync, or files manually added each and every time with every new update that was released. The approach of creating Pods works best for small and standalone repos, not for those that are far larger and rely on many others.

Going forward, we're hoping that Apple announces a better way to handle this within [Swift Package Manager](https://swift.org/package-manager/) soon for iOS, thus voiding the need for Pods to be created for Kitura (and removing the need for Cocoapods altogether).

**Comment by ddunn2** [comment]
@hactar You are correct, after a discussion we've decided to continue to go ahead with this. 

@ShihabMehboob and I are currently looking into it, and hope to have something in place soon!
